eth2near/eth2near-block-relay/index.js

```diff
@@ -40,8 +40,45 @@ function ethashproof (command, _callback) {
   )
 }
 
+// This function find the result in O(log delta) where delta is the difference between estimatedPosition and the result.
+// In particular if estimatedPosition is the correct value it will make two calls to predicate, so it will behave in O(1) in this case.
+async function binarySearchWithEstimate(limitLo, limitHi, estimatedPosition, predicate) {
+  let lo = limitLo;
+  let hi = limitHi;
+  let value = await predicate(estimatedPosition);
+
+  if (value) {
+    hi = estimatedPosition;
+    let step = 1;
+    while (hi - step > lo && await predicate(hi - step)) {
+      step *= 2;
+    }
+    hi -= Math.floor(step / 2);
+    lo = Math.max(lo, hi - step);
+  } else {
+    lo = estimatedPosition;
+    let step = 1;
+    while (lo + step < hi && !await predicate(lo + step)) {
+      step *= 2;
+    }
+    lo += Math.floor(step / 2);
+    hi = Math.min(hi, lo + step);
+  }
+
+
+  while (lo + 1 < hi) {
+    let mid = Math.floor((lo + hi) / 2);
+    if (await predicate(mid)) {
+      hi = mid;
+    } else {
+      lo = mid;
+    }
+  }
+  return hi;
+}
```

Comment:
I'm not sure if this is essential if we forbid Ropsten. This is most likely not required for Goerli and Mainnet.
Anyway, this is a good optimization, so it's anyway really great to have it!